### DISCUSSION TITLE
Is there a better way to achieve opening 3 tabs running a specific different command in each?

### DISCUSSION THREAD

**Discussion by 0xFloop** [wezterm/wezterm#4968] [Q&A] [answered]
### What Operating System(s) are you running on?

Linux Wayland

### Which Wayland compositor or X11 Window manager(s) are you using?

Mutter

### WezTerm version

20230712-072601-f4abf8fd

### Ask your question!

Hello and thanks for any help provided!

As the title suggests, I am wondering if anyone has thoughts on a better way to achieve the goal of opening a new wezterm window with 3 tabs each running a specific different command. Here is the current solution I am using.

```bash
#!/bin/bash

source .env

CWD='/home/unnamed/Desktop/code/fether'

wezterm start --cwd $CWD -- bash -c 'wezterm cli spawn --cwd '${CWD}' -- bash -c "wezterm cli spawn --cwd '${CWD}' -- pnpm anvil & pnpm ngrok;bash" & pnpm dev;bash'
```

I'm not very familiar with bash commands but this gets the job done however I'm sure triple nesting spawn commands isn't the best/right way to do it and would love some thoughts one improving it. I appreciate any and all help provided!

**Comment by wez** [comment]
Consider doing this from lua; see https://wezfurlong.org/wezterm/config/lua/gui-events/gui-startup.html

**Reply by 0xFloop** [reply]
Thank you for the response. 

I preferably want to be able to run the command from the command line. It is essentially a startup script for a project that runs a few servers.

The gui-startup event seems to not be suitable as i don't want this command to run every single time Wezterm is launched, please correct me if that is not the case. The Wezterm lua script does seem like a good solution if i can instead choose when to run it opposed to being only event based. I guess I could use a keybind but id prefer to use a set command line command. Thank you for the help!

**Reply by gegoune** [reply]
You can adapt that to run as a function triggered by key combination. I have few of those set up.

**Reply by 0xFloop** [reply]
Any thoughts on getting it to run from the command line?

**Reply by gegoune** [reply]
No, sorry, never had a need for that.

**Reply by edshamis** [reply]
Have a look https://wezfurlong.org/wezterm/config/lua/window-events/user-var-changed.html

**Reply by wez** [reply, marked as answer]
if you want to spawn three things from a bash script, you can write that in the bash script:

```bash
#!/bin/bash

CWD='/home/unnamed/Desktop/code/fether'

wezterm start --new-tab --cwd $CWD -- first program
wezterm start --new-tab --cwd $CWD -- second program
wezterm start --new-tab --cwd $CWD -- third program
```

(you need at least version `20240203-110809-5046fc22` of wezterm to use `wezterm start --new-tab`)

**Reply by 0xFloop** [reply]
This is definitely a much better solution thank you very much. However, upon updating I now get some 'Ignoring XDG_SESSION_TYPE' warnings on some commands. Also, wezterm now seems to be lagging after each command. IE if i run 'ls' it immediately lists out contents but takes about half a second to return to the bash input prompt

**Reply by 0xFloop** [reply]
Ope, sorry, seems as though these issues were due to multiple versions installed at once, all is good now. Thanks for all the help @gegoune @edshamis @wez !

**Reply by 0xFloop** [reply]
I also went a bit further and wrote some more to add correct titles to each new tab. Couldn't seem to find a way to do this with the --new-tab flag. Would be a nice feature to enable tab title in conjunction with the --new-tab flag 

```bash
#!/bin/bash

CWD='/home/unnamed/Desktop/code/fether'

wezterm start --new-tab --cwd $CWD -- pnpm dev 
wezterm start --new-tab --cwd $CWD -- pnpm anvil
wezterm start --new-tab --cwd $CWD -- pnpm ngrok
wezterm start --new-tab --cwd $CWD -- pnpm studio

OP=$(wezterm cli list | awk '{print $2}')
TAB_NAMES=("Studio" "Ngrok" "Anvil" "Dev")
TAB_IDS=()
SCRIPT_CREATED_TAB_IDS=()
for i in $OP; do 
	if [ "$i" != "TABID" ]
	then
		TAB_IDS+=($i)
	fi
done

for i in {1..4}; do 
	SCRIPT_CREATED_TAB_IDS+=(${TAB_IDS[$i * -1]})
done
TAB_NAME_INDEX=0
for i in "${SCRIPT_CREATED_TAB_IDS[@]}"; do
	wezterm cli set-tab-title --tab-id $i ${TAB_NAMES[$TAB_NAME_INDEX]}
	TAB_NAME_INDEX=$((TAB_NAME_INDEX+1))
done
```
